liquidation-monitor/src/agent.js

```diff
@@ -0,0 +1,383 @@
+    const findings = [];
+    const {
+      comptrollerContract,
+      oneInchContract,
+      accounts,
+      tokens,
+      borrow,
+      supply,
+    } = data;
+
+    // Initialize accounts. New accounts will get updated in the block section
+    data.newAccounts.forEach((newAccount) => { accounts[newAccount.toLowerCase()] = {}; });
+    data.totalNewAccounts += data.newAccounts.length;
+    data.newAccounts = [];
+
+    // Update Balances on zero health accounts
+    const filteredAccounts = [];
+    Object.keys(accounts).forEach((currentAccount) => {
+      if (accounts[currentAccount].health == null
+        || accounts[currentAccount].health === 0) {
+        filteredAccounts.push(currentAccount);
+        // Zero account balances
+        Object.keys(supply).forEach((currentToken) => {
+          supply[currentToken][currentAccount] = null;
+        });
+        Object.keys(borrow).forEach((currentToken) => {
+          borrow[currentToken][currentAccount] = null;
+        });
+      }
+    });
+
+    // Grab the assets in first, and make sure they are initialized
+    const foundTokens = [];
+    await Promise.all(filteredAccounts.map(async (currentAccount) => {
+      const assetsIn = await comptrollerContract.getAssetsIn(currentAccount);
+      accounts[currentAccount].assetsIn = assetsIn.map((asset) => {
+        const address = asset.toLowerCase();
+        foundTokens.push(address);
+        return address;
+      });
+    }));
+    // Use Array to Set to Array to remove duplicates
+    const filteredTokens = Array.from(new Set(foundTokens));
+
+    // Initialize token objects
+    await Promise.all(filteredTokens.map(async (token) => {
+      await verifyToken(data, token);
+    }));
+
+    // Grab token balances from on-chain
+    await Promise.all(filteredAccounts.map(async (currentAccount) => {
+      await Promise.all(accounts[currentAccount].assetsIn.map(async (currentToken) => {
+        const snapshot = await tokens[currentToken].contract.getAccountSnapshot(currentAccount);
+
+        // update the supply balance and token quantity
+        supply[currentToken][currentAccount] = BigNumber(snapshot[1].toString())
+          .dividedBy(tokens[currentToken].cTokenDecimalsMult);
+
+        // update the borrow balance
+        borrow[currentToken][currentAccount] = BigNumber(snapshot[2].toString());
+      }));
+    }));
+
+    // Update all token prices via 1inch
+    // Note: Object.entries does not work here because the nested objects cannot be retrieved.
+    await Promise.all(Object.keys(tokens).map(async (currentToken) => {
+      const price = await oneInchContract.getRateToEth(tokens[currentToken].underlying, 0);
+
+      // 1inch's getRateToEth uses 36 decimal places. The space is shared with the token's decimal.
+      //   So (36 - tokenDecimal) will yield the actual decimal for the rate.
+      const oneInchMult = BigNumber(10).pow(36 - tokens[currentToken].tokenDecimals);
```

Comment:
Based on the complexity of the 1inch contract math and attempting to compensate for it here, it would be better to select a simpler smart contract oracle.  Options include Chainlink Aggregators, as we used for Aave:
https://github.com/arbitraryexecution/forta-aave-suite/blob/main/compare-oracle-to-fallback/src/agent.js

Or Uniswap, which we used for Gamma (nee Visor): see our Confluence page on how to get the price ratio

Reply:
I had issues with my implementation of Chainlink so I removed it in an earlier commit. I don't see this any more complex than the `exchangeRateStored` math in the cTokens, but I can switch it to UNI. I felt that 1inch was more robust because it's price feed is pulled form UNI v1, v2, v3, Chainlink, Sushi and more.

Reply:
Clarified the 1inch scaling (with source) in commit b1fdb3a86d5ca1d9e57500c9c1a43f77d2d9b138